File: Question/31_Next_Permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        if(nums.size() <= 1) return;
        int i;
        int j;
        for(i = nums.size() - 1; i >=1; --i) {
            if(nums[i] > nums[i - 1]) { //begin to decrease
                for(j = nums.size() - 1; j >= i; --j) {
                    if(nums[j] > nums[i - 1])
                        break;
                }
                swap(nums[i - 1], nums[j]);
                reverse(nums.begin() + i, nums.end());
                return;
            }
        }
        reverse(nums.begin(), nums.end());
    }
};

File: Question_Java/31_Next_Permutation.java
class Solution {
    public void nextPermutation(int[] nums) {
        if (nums.length <= 1)
            return;
        for (int i = nums.length - 1; i > 0; --i) {
            if (nums[i] > nums[i - 1]) {
                int j = nums.length - 1;
                for (j = nums.length - 1; j >= i; --j) {
                    if (nums[j] > nums[i - 1])
                        break;
                }
                swap(nums, i - 1, j);
                reverse(nums, i, nums.length - 1);
                return;
            }
        }
        reverse(nums, 0, nums.length - 1);
    }

    private void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }

    private void reverse(int[] nums, int i, int j) {
        while (i < j) {
            swap(nums, i++, j--);
        }
    }
}